### DISCUSSION TITLE
Executing turbo tasks for staged files

### DISCUSSION THREAD

**Discussion by robert-melonn** [vercel/turborepo#9300] [Help] [answered]
### Summary

Hi guys, I'm working on migrating a monorepo to turborepo setup. I have been trying to find a good way to execute `turbo run format lint test` just for those workspaces that have been affected until precommit hook  (staged files). I'm using lefthook to achieve it but it has been impossible for me doing the turbo filter. I tried with `--affected` but is really equivalent to `--filter=[main...HEAD]`. I tried with `--filter=[HEAD]` but is taking non staged files too. 

Do you have any idea how you can handle this use case with git hooks ? can you give me an example pls?

### Additional information

PD: previous configuration used lint-staged and we had a `precommit ` in each workspace in this executed lint-stagedrc.js with `['npm run lint', 'npm run format', 'npm run test'] `. but I want to approach the turbo package graph to create these kinds of dependencies between them
_No response_

**Comment by bhunjadi** [comment, marked as answer]
We're having the same kind of issue after migrating from nx. 

I found this might work:
```bash
turbo ls --filter="...[--staged]"
# or 
turbo ls --filter="...[--cached]"
```

Because of `...` in filter, this will also include all workspaces that depend on the changed workspaces. Leave three dots out if this is not needed.
I came to this solution because it looks like turbo is using `git diff` behind the scenes. It feels a bit hacky, though.

If you'd want unstaged files, you could use `turbo ls --filter="[--]"` or `turbo ls --filter="...[--]"` if you want dependent workspaces.

Did you come up with a better solution?

**Reply by robert-melonn** [reply]
i think your answer is awesome, it could solve my problem. before this I made my implementation with lint-staged without using turbo

**Reply by bhunjadi** [reply]
Looks like this is no longer working in turbo@2.4.1 and possibly earlier. I was at 2.3.0 before and this one worked.
It seems they changed from git diff to git diff-tree https://github.com/vercel/turborepo/blob/032449ee539a34d1ff8d7cd37ee92596a66c9d07/crates/turborepo-scm/src/git.rs#L292 and the commands above are not longer working.
